README.md
# educandu

[![codecov](https://codecov.io/gh/educandu/educandu/branch/main/graph/badge.svg?token=SM7ANNBT3A)](https://codecov.io/gh/educandu/educandu)

The educandu framework

## Perequisites

 * Node 16 or higher
 * Docker
 * gulp installed globally (for conveninence): npm i -g gulp

 ## Configuration options

 | Option | Description | Type | Required |
 | --- | --- | --- | --- |
 | port | Port on which the project is run | `number`, mininum 1 | no, defaults to 80 |
 | mongoConnectionString | The URI for the project's MongoDB | `string` | yes |
 | skipMongoMigrations | Whether or not to run newly added MongoDB migration scripts on startup | `boolean` | no, defaults to `false` |
 | skipMongoChecks | Whether or not to run MongoDB checks resulting in ensuring DB collections on startup | `boolean` | no, defaults to `false` |
 | cdnEndpoint | The URL of the AWS-hosted CDN | `string` | yes |
 | cdnRegion | The region of the AWS-hosted CDN | `string` | yes |
 | cdnAccessKey | The access key of the AWS-hosted CDN | `string` | yes |
 | cdnSecretKey | The secret key of the AWS-hosted CDN | `string` | yes |
 | cdnBucketName | The name of the AWS S3 bucket storing the CDN data | `string` | yes |
 | cdnRootUrl | The root url of the CDN | `string` | yes |
 | sessionSecret | The unique ID of the user session | `string` | no, defaults to a generated unique id |
 | sessionDurationInMinutes | The validity of the user session in minutes | `number`, minumum 1 | no, defaults to 60 |
 | smtpOptions | The SMTP setup for sending emails to users upon registration or password reset | anything | yes |
 | emailSenderAddress | The email address from which emails are sent | `string` | yes |
 | publicFolders | The project-specific public folders that need to be accesible on the project domain | `array` of string | no |
 | resources | URLs to additional resource bundles, e.g. extra translations  | `array` of string  | no |
 | initialUser | The first user account, with admin role | `{ username, password, email }` or `null` | no |
 | exposeErrorDetails | Whether or not to expose details of thrown errors (e.g. stack trace) | `boolean` | no, defaults to `false` |
 | exportApiKey | The API key used for authorizing incoming requests for fetching data to be imported into another system | `string` | no |
 | importSources | The system from which data can be imported | `[{ name, hostName, allowUnsecure, apiKey }]` | no, if not specified then `allowUnsecure` defaults to `false` |
 | disabledFeatures | A list of names of disabled features | `string[]` | no |
 | taskProcessing | Task processing setup | `{ isEnabled, idlePollIntervalInMs, maxAttempts }` | no, defaults to `{ isEnabled: false, idlePollIntervalInMs: 5000, maxAttempts: 3 }` |

## How to use

~~~
$ yarn add @educandu/educandu
~~~

Use it in code as follows:

~~~
import educandu from '@educandu/educandu';

educandu({
  port: 3000,
  mongoConnectionString: 'mongodb://root:rootpw@localhost:27017/dev-educandu-db?replicaSet=educandurs&authSource=admin',
  skipMongoMigrations: process.env.TEST_APP_SKIP_MONGO_MIGRATIONS === true.toString(),
  skipMongoChecks: process.env.TEST_APP_SKIP_MONGO_CHECKS === true.toString(),
  cdnEndpoint: 'http://localhost:9000',
  cdnRegion: 'eu-central-1',
  cdnAccessKey: 'UVDXF41PYEAX0PXD8826',
  cdnSecretKey: 'SXtajmM3uahrQ1ALECh3Z3iKT76s2s5GBJlbQMZx',
  cdnBucketName: 'dev-educandu-cdn',
  cdnRootUrl: 'http://localhost:9000/dev-educandu-cdn',
  sessionSecret: 'd4340515fa834498b3ab1aba1e4d9013',
  sessionDurationInMinutes: 60,
  emailSenderAddress: 'educandu-test-app@test.com',
  smtpOptions: 'smtp://localhost:8025/?ignoreTLS=true',
  publicFolders: ['./test-app/dist', './test-app/static'].map(x => path.resolve(x)),
  resources: ['./test-app/resource-overrides.json'].map(x => path.resolve(x)),
  initialUser: {
    username: 'test',
    password: 'test',
    email: 'test@test.com'
  },
  exposeErrorDetails: true,
  exportApiKey: 'GSD54GDFgDGesdfs4'
  importSources: [{
    name: 'otherSystem',
    hostName: 'othersystem.com',
    apiKey: 'FSERFSF56Ggsdfg6FGED'
  }]
});
~~~

This should run the entire application and provide you with an admin user as defined on the initialUser setting.

## How to override the styles

In order to override the styles provided by educandu you need to import the main.less of the educandu project and override the less variables provided. An example override implementation is provided by the test-app in the 
main.less file. We will keep this file updated so any user will know what can be overriden. Alternatively you can go 
to the global-variables.less in educandu and consult the list there.

## How to run and develop locally

~~~
$ yarn
$ gulp
~~~

This will build and start up the TestApp (in watch mode), which is set up to use educandu.

By default the application requires that the following ports are available to be taken:
  * 3000: the application itself
  * 8000: maildev server, can be used for debugging emails that would be sent to the users (head to http://localhost:8000 when the application is running)
  * 8025: smpt server port
  * 9000: minio docker image (local CDN used for testing)
  * 21017: mongodb docker image port

The ports can be changed in the gulp file and must be changed in the gulpfile.js and need to be reflected in the test-app/index.js.

The gulpfile has a number of useful tasks which can be run with "gulp taskName", here are some more widely used:
 * test: runs all tests
 * testChanged: runs tests that were affected by the current modifications
 * lint: runs lint
 * down: stops all the containers and deletes them


## License

Educandu is released under the MIT License. See the bundled LICENSE file for details.


CHANGELOG.md
# Changelog

## [Unreleased](https://github.com/educandu/educandu/tree/HEAD)

[Full Changelog](https://github.com/educandu/educandu/compare/0.1.0...HEAD)

**Merged pull requests:**

- [EDU-95](https://educandu.atlassian.net/browse/EDU-95) Allow null initial user [\#45](https://github.com/educandu/educandu/pull/45) ([laurentiu-ilici](https://github.com/laurentiu-ilici))
- [EDU-117](https://educandu.atlassian.net/browse/EDU-117) Fix showing revisions slider for non admin users [\#44](https://github.com/educandu/educandu/pull/44) ([adrianaluchian](https://github.com/adrianaluchian))
- [EDU-116](https://educandu.atlassian.net/browse/EDU-116) Pin mongodb at version 4.1.4 [\#42](https://github.com/educandu/educandu/pull/42) ([ahelmberger](https://github.com/ahelmberger))
- [EDU-115](https://educandu.atlassian.net/browse/EDU-115) Remove `env` from client config [\#41](https://github.com/educandu/educandu/pull/41) ([ahelmberger](https://github.com/ahelmberger))
- [EDU-100](https://educandu.atlassian.net/browse/EDU-100) Fix migration bug [\#40](https://github.com/educandu/educandu/pull/40) ([laurentiu-ilici](https://github.com/laurentiu-ilici))
- [EDU-106](https://educandu.atlassian.net/browse/EDU-106) Enable API key authentication [\#38](https://github.com/educandu/educandu/pull/38) ([ahelmberger](https://github.com/ahelmberger))
- [EDU-88](https://educandu.atlassian.net/browse/EDU-88) adaptible styles [\#37](https://github.com/educandu/educandu/pull/37) ([laurentiu-ilici](https://github.com/laurentiu-ilici))
- [EDU-108](https://educandu.atlassian.net/browse/EDU-108) Fix duplicate key issue [\#36](https://github.com/educandu/educandu/pull/36) ([adrianaluchian](https://github.com/adrianaluchian))
- [EDU-100](https://educandu.atlassian.net/browse/EDU-100)-migrate-cdn-resources [\#35](https://github.com/educandu/educandu/pull/35) ([laurentiu-ilici](https://github.com/laurentiu-ilici))
- [EDU-99](https://educandu.atlassian.net/browse/EDU-99) Collect CDN resources [\#34](https://github.com/educandu/educandu/pull/34) ([ahelmberger](https://github.com/ahelmberger))
- [EDU-103](https://educandu.atlassian.net/browse/EDU-103) Do a full page refresh \(without using the cache\) when home page is loaded from cache [\#33](https://github.com/educandu/educandu/pull/33) ([adrianaluchian](https://github.com/adrianaluchian))
- [EDU-102](https://educandu.atlassian.net/browse/EDU-102) Add import config [\#32](https://github.com/educandu/educandu/pull/32) ([adrianaluchian](https://github.com/adrianaluchian))
- [EDU-98](https://educandu.atlassian.net/browse/EDU-98) Add endpoint for importable documents [\#31](https://github.com/educandu/educandu/pull/31) ([ahelmberger](https://github.com/ahelmberger))
- [EDU-77](https://educandu.atlassian.net/browse/EDU-77) Support latest node versions [\#30](https://github.com/educandu/educandu/pull/30) ([laurentiu-ilici](https://github.com/laurentiu-ilici))
- [EDU-97](https://educandu.atlassian.net/browse/EDU-97) Add origin to documents [\#29](https://github.com/educandu/educandu/pull/29) ([adrianaluchian](https://github.com/adrianaluchian))
- [EDU-96](https://educandu.atlassian.net/browse/EDU-96) Prefill filters on search results [\#28](https://github.com/educandu/educandu/pull/28) ([laurentiu-ilici](https://github.com/laurentiu-ilici))
- [EDU-93](https://educandu.atlassian.net/browse/EDU-93) change provider name [\#27](https://github.com/educandu/educandu/pull/27) ([adrianaluchian](https://github.com/adrianaluchian))
- [EDU-58](https://educandu.atlassian.net/browse/EDU-58) Hide archived documents on search [\#26](https://github.com/educandu/educandu/pull/26) ([laurentiu-ilici](https://github.com/laurentiu-ilici))
- [EDU-74](https://educandu.atlassian.net/browse/EDU-74) Update readme [\#25](https://github.com/educandu/educandu/pull/25) ([laurentiu-ilici](https://github.com/laurentiu-ilici))
- [EDU-85](https://educandu.atlassian.net/browse/EDU-85) Remove aurelia-dependency-injection [\#24](https://github.com/educandu/educandu/pull/24) ([ahelmberger](https://github.com/ahelmberger))
- [EDU-64](https://educandu.atlassian.net/browse/EDU-64) Make dependencies more flexible [\#23](https://github.com/educandu/educandu/pull/23) ([ahelmberger](https://github.com/ahelmberger))
- [EDU-84](https://educandu.atlassian.net/browse/EDU-84) Measure bundle size in test app [\#22](https://github.com/educandu/educandu/pull/22) ([ahelmberger](https://github.com/ahelmberger))
- [EDU-72](https://educandu.atlassian.net/browse/EDU-72) replace momentjs with date fns [\#21](https://github.com/educandu/educandu/pull/21) ([laurentiu-ilici](https://github.com/laurentiu-ilici))
- [EDU-71](https://educandu.atlassian.net/browse/EDU-71) Add document language to the search results [\#20](https://github.com/educandu/educandu/pull/20) ([laurentiu-ilici](https://github.com/laurentiu-ilici))
- [EDU-68](https://educandu.atlassian.net/browse/EDU-68) Make sender email configurable [\#19](https://github.com/educandu/educandu/pull/19) ([ahelmberger](https://github.com/ahelmberger))
- [EDU-60](https://educandu.atlassian.net/browse/EDU-60) correct document cloning [\#18](https://github.com/educandu/educandu/pull/18) ([adrianaluchian](https://github.com/adrianaluchian))
- [EDU-63](https://educandu.atlassian.net/browse/EDU-63) Correct server configs [\#17](https://github.com/educandu/educandu/pull/17) ([laurentiu-ilici](https://github.com/laurentiu-ilici))
- [EDU-59](https://educandu.atlassian.net/browse/EDU-59) Fix terms and conditions validation on user registration [\#16](https://github.com/educandu/educandu/pull/16) ([laurentiu-ilici](https://github.com/laurentiu-ilici))
- [EDU-61](https://educandu.atlassian.net/browse/EDU-61) Fix firing too many events [\#15](https://github.com/educandu/educandu/pull/15) ([laurentiu-ilici](https://github.com/laurentiu-ilici))

gulpfile.js
/* eslint-disable no-console, no-process-env, require-atomic-updates */
import url from 'url';
import del from 'del';
import path from 'path';
import gulp from 'gulp';
import glob from 'glob';
import yaml from 'yaml';
import execa from 'execa';
import less from 'gulp-less';
import csso from 'gulp-csso';
import gulpif from 'gulp-if';
import esbuild from 'esbuild';
import eslint from 'gulp-eslint';
import { promisify } from 'util';
import plumber from 'gulp-plumber';
import superagent from 'superagent';
import { promises as fs } from 'fs';
import Graceful from 'node-graceful';
import { spawn } from 'child_process';
import { Docker } from 'docker-cli-js';
import sourcemaps from 'gulp-sourcemaps';
import { Client as MinioClient } from 'minio';
import LessAutoprefix from 'less-plugin-autoprefix';

const ROOT_DIR = path.dirname(url.fileURLToPath(import.meta.url));

const TEST_MAILDEV_IMAGE = 'maildev/maildev:1.1.0';
const TEST_MAILDEV_CONTAINER_NAME = 'maildev';

const TEST_MONGO_IMAGE = 'bitnami/mongodb:4.2.17-debian-10-r23';
const TEST_MONGO_CONTAINER_NAME = 'mongo';

const TEST_MINIO_IMAGE = 'bitnami/minio:2020.12.18';
const TEST_MINIO_CONTAINER_NAME = 'minio';

const MINIO_ACCESS_KEY = 'UVDXF41PYEAX0PXD8826';
const MINIO_SECRET_KEY = 'SXtajmM3uahrQ1ALECh3Z3iKT76s2s5GBJlbQMZx';

const optimize = (process.argv[2] || '').startsWith('ci') || process.argv.includes('--optimize');

const supportedLanguages = ['en', 'de'];

let testAppServer = null;
let testAppBuildResult = null;
const containerCommandTimeoutMs = 2000;

const delay = ms => new Promise(resolve => {
  setTimeout(resolve, ms);
});

Graceful.on('exit', () => {
  testAppBuildResult?.rebuild?.dispose();
  return new Promise(resolve => {
    if (testAppServer) {
      testAppServer.once('exit', () => resolve());
      testAppServer.kill();
    } else {
      resolve();
    }
  });
});

const kebabToCamel = str => str.replace(/-[a-z0-9]/g, c => c.toUpperCase()).replace(/-/g, '');

const ensureContainerRunning = async ({ containerName, runArgs, afterRun = () => Promise.resolve() }) => {
  const docker = new Docker();
  const data = await docker.command('ps -a');
  const container = data.containerList.find(c => c.names === containerName);
  if (!container) {
    await docker.command(`run --name ${containerName} ${runArgs}`);
    await delay(containerCommandTimeoutMs);
    await afterRun();
  } else if (!container.status.startsWith('Up')) {
    await docker.command(`restart ${containerName}`);
    await delay(containerCommandTimeoutMs);
  }
};

const ensureContainerRemoved = async ({ containerName }) => {
  const docker = new Docker();
  try {
    await docker.command(`rm -f ${containerName}`);
    await delay(containerCommandTimeoutMs);
  } catch (err) {
    if (!err.toString().includes('No such container')) {
      throw err;
    }
  }
};

async function ensureBucketExists() {
  const region = 'eu-central-1';
  const bucketName = 'dev-educandu-cdn';

  const minioClient = new MinioClient({
    endPoint: 'localhost',
    port: 9000,
    useSSL: false,
    region,
    accessKey: MINIO_ACCESS_KEY,
    secretKey: MINIO_SECRET_KEY
  });

  const bucketPolicy = {
    Version: '2012-10-17',
    Statement: [
      {
        Sid: 'PublicReadGetObject',
        Effect: 'Allow',
        Principal: '*',
        Action: 's3:GetObject',
        Resource: `arn:aws:s3:::${bucketName}/*`
      }
    ]
  };

  const buckets = await minioClient.listBuckets();

  if (!buckets.find(x => x.name === 'dev-educandu-cdn')) {
    await minioClient.makeBucket(bucketName, region);
    await minioClient.setBucketPolicy(bucketName, JSON.stringify(bucketPolicy));
  }
}

function runJest(...flags) {
  return execa(process.execPath, [
    '--experimental-vm-modules',
    `${ROOT_DIR}/node_modules/jest/bin/jest.js`,
    ...flags.map(flag => `--${flag}`)
  ], { stdio: 'inherit' });
}

const downloadCountryList = async lang => {
  const res = await superagent.get(`https://raw.githubusercontent.com/umpirsky/country-list/master/data/${lang}/country.json`);
  await fs.writeFile(`./src/data/country-names/${lang}.json`, JSON.stringify(JSON.parse(res.text), null, 2), 'utf8');
};

export async function clean() {
  await del(['.tmp', 'dist', 'coverage', 'reports', 'test-app/dist']);
}

export function lint() {
  return gulp.src(['*.js', 'src/**/*.js', 'scripts/**/*.js'], { base: './' })
    .pipe(eslint())
    .pipe(eslint.format())
    .pipe(gulpif(!testAppServer, eslint.failAfterError()));
}

export function fix() {
  return gulp.src(['*.js', 'src/**/*.js', 'scripts/**/*.js'], { base: './' })
    .pipe(eslint({ fix: true }))
    .pipe(eslint.format())
    .pipe(gulpif(file => file.eslint?.fixed, gulp.dest('./')))
    .pipe(eslint.failAfterError());
}

export function test() {
  return runJest('coverage', 'runInBand');
}

export function testChanged() {
  return runJest('onlyChanged');
}

export function testWatch() {
  return runJest('watch');
}

export async function buildJs() {
  const jsFiles = await promisify(glob)('src/**/*.js', { ignore: 'src/**/*.spec.js' });
  Promise.all(jsFiles.map(jsFile => {
    return esbuild.build({
      entryPoints: [jsFile],
      target: ['esnext'],
      format: 'esm',
      loader: { '.js': 'jsx' },
      sourcemap: true,
      sourcesContent: true,
      outfile: path.resolve('./dist', path.relative('src', jsFile))
    });
  }));
}

export function copyToDist() {
  return gulp.src(['src/**', '!src/**/*.{js,yml}'], { base: 'src' })
    .pipe(gulp.dest('dist'));
}

export const build = gulp.parallel(copyToDist, buildJs);

export function buildTestAppCss() {
  return gulp.src('test-app/main.less')
    .pipe(gulpif(!!testAppServer, plumber()))
    .pipe(sourcemaps.init())
    .pipe(less({ javascriptEnabled: true, plugins: [new LessAutoprefix({ browsers: ['last 2 versions'] })] }))
    .pipe(gulpif(optimize, csso()))
    .pipe(sourcemaps.write('.'))
    .pipe(gulp.dest('test-app/dist'));
}

export async function buildTestAppJs() {
  if (testAppBuildResult && testAppBuildResult.rebuild) {
    await testAppBuildResult.rebuild();
  } else {
    testAppBuildResult = await esbuild.build({
      entryPoints: await promisify(glob)('./test-app/bundles/*.js'),
      target: ['esnext', 'chrome95', 'firefox93', 'safari15', 'edge95'],
      format: 'esm',
      bundle: true,
      splitting: true,
      incremental: !!testAppServer,
      minify: optimize,
      loader: { '.js': 'jsx' },
      inject: ['./test-app/polyfills.js'],
      metafile: optimize,
      sourcemap: true,
      sourcesContent: true,
      outdir: './test-app/dist'
    });

    if (testAppBuildResult.metafile) {
      const bundles = Object.entries(testAppBuildResult.metafile.outputs)
        .map(([name, { bytes }]) => ({ name, bytes }))
        .filter(x => x.name.endsWith('.js'));

      const formatBytes = bytes => `${(bytes / 1000).toFixed(2)} kB`;
      bundles.forEach(({ name, bytes }) => console.log(`${name}: ${formatBytes(bytes)}`));
      console.log(`TOTAL: ${formatBytes(bundles.reduce((sum, { bytes }) => sum + bytes, 0))}`);

      await fs.writeFile('./test-app/dist/meta.json', JSON.stringify(testAppBuildResult.metafile, null, 2), 'utf8');
    }
  }
}

export async function buildTranslations() {
  const filePaths = await promisify(glob)('./src/**/*.yml');

  const bundleGroups = await Promise.all(filePaths.map(async filePath => {
    const namespace = kebabToCamel(path.basename(filePath, '.yml'));
    const content = await fs.readFile(filePath, 'utf8');
    const resources = yaml.parse(content);

    if (!resources) {
      return [];
    }

    const bundlesByLanguage = {};
    const resourceKeys = Object.keys(resources);

    resourceKeys.forEach(resourceKey => {
      const languages = Object.keys(resources[resourceKey]);
      languages.forEach(language => {
        let languageBundle = bundlesByLanguage[language];
        if (!languageBundle) {
          languageBundle = {
            namespace,
            language,
            resources: {}
          };
          bundlesByLanguage[language] = languageBundle;
        }
        languageBundle.resources[resourceKey] = resources[resourceKey][language];
      });
    });

    return Object.values(bundlesByLanguage);
  }));

  const result = bundleGroups.flatMap(x => x);

  await fs.writeFile('./src/resources/resources.json', JSON.stringify(result, null, 2), 'utf8');
}

export const buildTestApp = gulp.parallel(buildTestAppCss, buildTranslations, buildTestAppJs);

export async function countriesUpdate() {
  await Promise.all(supportedLanguages.map(downloadCountryList));
}

export async function maildevUp() {
  await ensureContainerRunning({
    containerName: TEST_MAILDEV_CONTAINER_NAME,
    runArgs: `-d -p 8000:80 -p 8025:25 ${TEST_MAILDEV_IMAGE}`
  });
}

export async function maildevDown() {
  await ensureContainerRemoved({
    containerName: TEST_MAILDEV_CONTAINER_NAME
  });
}

export const maildevReset = gulp.series(maildevDown, maildevUp);

export async function mongoUp() {
  await ensureContainerRunning({
    containerName: TEST_MONGO_CONTAINER_NAME,
    runArgs: [
      '-d',
      '-p 27017:27017',
      '-e MONGODB_ROOT_USER=root',
      '-e MONGODB_ROOT_PASSWORD=rootpw',
      '-e MONGODB_REPLICA_SET_KEY=educandurs',
      '-e MONGODB_REPLICA_SET_NAME=educandurs',
      '-e MONGODB_REPLICA_SET_MODE=primary',
      '-e MONGODB_ADVERTISED_HOSTNAME=localhost',
      TEST_MONGO_IMAGE
    ].join(' ')
  });
}

export async function mongoDown() {
  await ensureContainerRemoved({
    containerName: TEST_MONGO_CONTAINER_NAME
  });
}

export const mongoReset = gulp.series(mongoDown, mongoUp);

export async function minioUp() {
  await ensureContainerRunning({
    containerName: TEST_MINIO_CONTAINER_NAME,
    runArgs: [
      '-d',
      '-p 9000:9000',
      `-e MINIO_ACCESS_KEY=${MINIO_ACCESS_KEY}`,
      `-e MINIO_SECRET_KEY=${MINIO_SECRET_KEY}`,
      '-e MINIO_BROWSER=on',
      TEST_MINIO_IMAGE
    ].join(' '),
    afterRun: ensureBucketExists
  });
}

export async function minioDown() {
  await ensureContainerRemoved({
    containerName: TEST_MINIO_CONTAINER_NAME
  });
}

export const minioReset = gulp.series(minioDown, minioUp);

function startTestApp({ skipMigrationsAndChecks }) {
  testAppServer = spawn(
    process.execPath,
    [
      '--experimental-json-modules',
      '--experimental-loader',
      '@educandu/node-jsx-loader',
      '--enable-source-maps',
      'test-app/index.js'
    ],
    {
      env: {
        ...process.env,
        NODE_ENV: 'development',
        TEST_APP_SKIP_MONGO_MIGRATIONS: (!!skipMigrationsAndChecks).toString(),
        TEST_APP_SKIP_MONGO_CHECKS: (!!skipMigrationsAndChecks).toString()
      },
      stdio: 'inherit',
      detached: true
    }
  );
  testAppServer.once('exit', () => {
    testAppServer = null;
  });
}

export function startServer(done) {
  startTestApp({ skipMigrationsAndChecks: false });
  done();
}

export function restartServer(done) {
  if (testAppServer) {
    testAppServer.once('exit', () => {
      startTestApp({ skipMigrationsAndChecks: true });
      done();
    });
    testAppServer.kill();
  } else {
    startTestApp({ skipMigrationsAndChecks: true });
    done();
  }
}

export function runGithubChangelogGenerator() {
  return execa.command('github_changelog_generator -u educandu -p educandu --no-author', { stdio: 'inherit' });
}

export async function generateJiraLinks() {
  const changeLog = await fs.readFile('CHANGELOG.md', 'utf8');
  const replacedChangeLog = changeLog.replace(/EDU-\d+/g, num => `[${num}](https://educandu.atlassian.net/browse/${num})`);
  await fs.writeFile('CHANGELOG.md', replacedChangeLog, 'utf8');
}

export const generateChangelog = gulp.series(runGithubChangelogGenerator, generateJiraLinks);

export const up = gulp.series(mongoUp, minioUp, maildevUp);

export const down = gulp.parallel(mongoDown, minioDown, maildevDown);
export const serve = gulp.series(gulp.parallel(up, build), buildTestApp, startServer);

export const ci = gulp.series(clean, lint, test, build, generateChangelog);

export function setupWatchers(done) {
  gulp.watch(['src/**/*.{js,json}', 'test-app/**/*.{js,json}', '!test-app/dist/**'], gulp.series(buildTestAppJs, restartServer));
  gulp.watch(['src/**/*.less', 'test-app/**/*.less'], gulp.series(copyToDist, buildTestAppCss));
  gulp.watch(['src/**/*.yml'], buildTranslations);
  done();
}

export const startWatch = gulp.series(serve, setupWatchers);

export default startWatch;
